feat(service): shows attached DPPs

kleinfreund opened the pull request:
## Changes

Shows the data plane proxies attached to a service in the form of the same data plane proxy table used in the DPP list view. This association is based on the `*/dataplanes+insights` endpoint’s new `tag` query parameter allowing us to get DPPs for a service by querying for its tag.

Uses tag list component for data plane details view.

Resolves #463. Resolves #492.

Signed-off-by: Philipp Rudloff <philipp.rudloff@konghq.com>

## Screenshots

<details><summary>Before:</summary>

![Screenshot from 2022-12-06 09-49-12](https://user-images.githubusercontent.com/5774638/205864073-b572439a-a81c-4abe-84e1-22a8aae93a82.png)

</details>

<details><summary>After:</summary>

![image](https://user-images.githubusercontent.com/5774638/206151703-31d5237b-e5a7-4205-8b2c-521d3f3df10f.png)

</details>

kleinfreund commented:
> If I click [Create Dataplane Proxy] I get these errors in console:

@johncowen Good catch. Since we trigger data loading when route parameter change, we need to protect against the case of this happening when navigating away from this view. Fixed now.

> and the [Refresh] button doesn't seem to do anything, but I might be mistaken, maybe I'm doing something wrong?

What do you expect to happen here? The refresh button also won’t typically do anything in the other views because our mock data doesn’t change. All it does is trigger a component’s loading routine.

johncowen commented:
> What do you expect to happen here? The refresh button also won’t typically do anything in the other views because our mock data doesn’t change. All it does is trigger a component’s loading routine.

I would expect it to make a HTTP request to refresh the data. It seems to make 2 requests, but not totally sure as they are the same 2 requests that it does when it polls something, and if you wait a bit before pushing that button it makes those requests anyway, so hard to tell if its the 2 second polling on those 2 http requests or not. One thing I noticed is if you hit the refresh button another time, then definitely doesn't do anything, no matter how many times I push the button.

Playing with it again, depending on when I click the button, I can't tell whether it does something or not

kleinfreund commented:
> Tags when showing kuma.io/service link to: https://deploy-preview-479--kuma-gui.netlify.app/mesh/default/services?ns=backend can we have them link to https://deploy-preview-479--kuma-gui.netlify.app/mesh/default/services/backend?name=<theDppThat'sSelectedIfThere'sOne> ? If this is too complicated please just open an issue and don't block this PR on this

@lahabana This is unfortunately a bit more difficult because we don’t know if a service tag represents a `ServiceInsight` or an `ExternalService`. Depending on that, we need to select the dedicated view. That’s necessary because we need to query a different endpoint to get a service’s data. I’ll file an issue for this because I would also prefer this to link to the detail view.

> Also opened the slightly related: https://github.com/kumahq/kuma-gui/issues/492

Addressed this as part of this because it’s very simple.

> The line break in data plane proxy is weird:

Good catch. Fixed that, too.

lahabana commented:
> @lahabana This is unfortunately a bit more difficult because we don’t know if a service tag represents a ServiceInsight or an ExternalService. Depending on that, we need to select the dedicated view. That’s necessary because we need to query a different endpoint to get a service’s data. I’ll file an issue for this because I would also prefer this to link to the detail view.

Isn't it an actual service in most cases (like in the DPP view it can't be not a regular service as there's a DPP for it).

lahabana commented:
Actually the external-service view seems to display a DPP list which is not possible: https://deploy-preview-479--kuma-gui.netlify.app/mesh/default/external-services/httpbin?name=backend

I don't know if this is a mock data bug or something else

kleinfreund commented:
> Isn't it an actual service in most cases (like in the DPP view it can't be not a regular service as there's a DPP for it).

Presumably, but for that one time when it isn’t, I won’t know before making the request to either endpoint, right?

lahabana commented:
> Presumably, but for that one time when it isn’t, I won’t know before making the request to either endpoint, right?

note: thinking out loud here :)

Why have we got different URLs btw? Are these different components? Should the GUI for either service be: https://deploy-preview-479--kuma-gui.netlify.app/mesh/default/services/httpbin?name=backend

And depending on the service insight result we pick the right data?

johncowen commented:
Reading the above sounds like this is feature complete/aesthetically decided on, I'll take a pass at the code etc if so, oh there's a conflict, ping me when thats sorted and I'll take a look

kleinfreund commented:
@johncowen I’ve resolved the merge conflicts now.

kleinfreund commented:
I’ve noticed an issue with external services in the list view that I’ve now fixed (https://github.com/kumahq/kuma-gui/pull/479/commits/95170bef9df1d8930e9eb8d545413e0a2c194182).

johncowen commented on src/app/services/views/ServiceDetailView.vue:
```diff
@@ -0,0 +1,94 @@
+  // Don’t trigger a load when the user is navigating to another route.
+  if (route.name !== 'service-insights-detail-view') {
+    return
+  }
```
Bit of a general question: Is this a typical Vue thing that we need to remember to add these guards all the time in `watch`es?

johncowen commented on src/store/storeConfig.ts:
```diff
@@ -284,8 +288,13 @@ export const storeConfig: StoreOptions<State> = {
+      } else {
+        ClientStorage.remove('selectedMesh')
+      }
+
```
Did this happen in another PR also?

kleinfreund replied:
Oh yeah, true, I must’ve applied that to both PRs while I was looking at it. I’m removing it.

johncowen reviewed (commented):
I started taking a look at this, but then I started seeing bits that are also in another PR, or did I misunderstand whats going on here? This PR is super big on changes, so if you could take those extra bits out that would be great - if not let me know and I'll take a look anyway.

kleinfreund commented on src/app/notification-manager/components/NotificationManager.vue:
```diff
@@ -84,8 +85,16 @@ const hasAnyAction = computed(() => {
 function closeAlert(): void {
   isShowingAlert.value = false
+
+  ClientStorage.set('hideCheckMeshAlert', 'yes')
```
Oh yeah, true, I must’ve applied that to both PRs while I was looking at it. I’m removing it.

kleinfreund replied:
Oh, actually, this is different from https://github.com/kumahq/kuma-gui/pull/479#discussion_r1044353036.

It’s such a small change that I’d rather just leave it in here. Its logic is only affecting this file and is of no greater consequence than to persist the dismissal of the "check mesh" alert.

johncowen replied:
https://github.com/kumahq/kuma-gui/pull/466#discussion_r1042000898

> I’m not sure I understand this argument. All this data is runtime state and while it might never change that doesn’t mean it shouldn’t be stored in a Vuex store.

Would you count this as runtime state? You had a very strong opinion that even static, non-changing variables should be stored/retrieved via Vuex. Considering our application is almost exclusively read-only (the source of truth being stored on the backend), and this is possibly one of the very few times we are actually modifying application state, what's your opinion on whether this particular thing should be going via Vuex?

johncowen replied:
Well I dunno, I suppose thats what I'm asking you if that's your point/argument. What's the advantage of not doing it here? This is application state that changes during the application lifecycle, not something that is completely static.

kleinfreund replied:
Sure, yes, but this data here is not needed anywhere else. It doesn’t need to be _stored_.

johncowen replied:
Fair point, so does that mean:

> I’m not sure I understand this argument. All this data is runtime state and while it might never change that doesn’t mean it shouldn’t be stored in a Vuex store.

Should be more like:

> I’m not sure I understand this argument. All this data is runtime state (apart from when its not needed anywhere else) and while it might never change that doesn’t mean it shouldn’t be stored in a Vuex store.

johncowen replied:
What are your thoughts on using something like Pinia then:

https://pinia.vuejs.org/
> Pinia: The Vue Store that you will enjoy using
> Compared to Vuex, Pinia provides a simpler API with less ceremony, offers Composition-API-style APIs

Looks to me like a lot of Pinia usage within components looks like:

```javascript
import { useSomething, useSomethingElse } from  '@/somewhere'
const something = useSomething();
const somethingElse = useSomethingElse();
```

`something` and `somethingElse` then look and act almost just like regular JS, and they're smaller things, and can be easily and individually be swapped in and out for something different if necessary.

As an aside: I think all in all I'm generally cautious of having this great big single store thing to maintain that we have to keep in sync with the actual source of truth (the backend). It's kinda fine for SPAs that do stuff, like saving/editing and maintaining a source of truth ready to post back somewhere else. It's not like we need to change any extended state/data in our application as yet (or maybe ever?), its just read-only.

I know we can't just rip out existing infra, just wondered what your thoughts were in an ideal world, or even a longer term thing for the app (not saying we need to change anything just good to discuss and hear what you have to say)

Let's chat tomorrow.

kleinfreund replied:
I'm happy with swapping out Vuex with Pinia. I hope for an overall better typing landscape which is my biggest gripe with Vuex.

Generally, I see a big advantage when using a centralized store: you can add, with relative ease, client-side caching strategies. For example, consider that API calls aren’t directly invoked from our API wrapper (i.e. `kumaApi`) but via a Vuex store action. Such an action could first check for some additional store-kept flag that tells it whether this data was fetched before and opt not to trigger a new, possibly redundant API call.

I’ve implemented something like this before which roughly worked like this:

* a regular store action is available for (1) triggering the API wrapper method and (2) storing its result in the appropriate place
* a lazy variant of that store action is available which triggers the regular variant but only when a condition is fulfilled. In the strategy I implemented, we stored the promise returned by the API wrapper method in the store. This enables the lazy action to check for whether (1) it is set (i.e. the signal to not fetch again) and (2) whether it is fulfilled (used as a way to avoid triggering several parallel asynchronous fetches).

On top of this approach, one could add a timer that clears the stored signal (e.g. the promise) after a certain while which would be the cache busting part. Of course certain pages/actions would just simply trigger the non-lazy action to avoid this caching (e.g. the reload buttons we have in some places).

---

The synchronization of data question to me is only partially related to the question of whether one has a central store like a Vuex store. Either with or without such a store, one needs to implement a strategy for ensuring the data is as accurate as it needs to do while simultaneously trying to avoid to constantly fetch data which would put unnecessary load on the API server.

johncowen replied:
I think I mostly agree with the majority of this 👍 , but obviously for things that change and that we should keep listening (or checking) for changes to.

kleinfreund commented on src/app/common/DataOverview.vue:
```diff
@@ -63,8 +63,8 @@
-              v-if="typeof rowValue !== 'boolean'"
+            <StatusBadge
+              v-if="rowValue !== null"
               :status="rowValue"
```
The handling for statuses like these was a bit all over the place which I’m addressing with turning the EntityStatus component into the StatusBadge component.

I’ve checked all the uses of the component and they now either have a status object that StatusBadge knows how to handle or they’re `null` in which case we either hide it completely or show a `—` (though that’s handled outside of that component).

I do recall that we used to have an instance where `false` indicated some sort of "not available", but I don’t recall where that was and looking now, I also don’t find any instance of the `status` simply being `false`.

kleinfreund commented:
> I started taking a look at this, but then I started seeing bits that are also in another PR, or did I misunderstand whats going on here? This PR is super big on changes, so if you could take those extra bits out that would be great - if not let me know and I'll take a look anyway.

@johncowen It’s true that this is a big PR, but the big majority of changes are in fact part of the core of the PR.

The task involves two of our bigger views: the service detail view and the data plane list view. The latter because we now re-use the bigger part of it (its table and sidebar) in another component which required some heavier rewiring. This changeset also necessarily involves the service list view which uses a component that’s also used by the service detail view which needed a different behavior.

So what I could theoretically move out is the work on the StatusBadge which while related to a couple of these components _could_ happen in another PR and the change to the kumaApi types (which are only partially unrelated, see `DataPlaneOverviewParameters`). Let me know if you want me to move out the StatusBadge changes.

johncowen commented:
> @johncowen It’s true that this is a big PR, but the big majority of changes are in fact part of the core of the PR.

Yeah I gotcha, I got the impression that this needed to be fairly big - it was more all the unrelated things in here, which as this was gonna be a fair size anyway it kinda made sense to move out, which I think you've done now anyhow.

> So what I could theoretically move out is the work on the StatusBadge which while related to a couple of these components could happen in another PR and the change to the kumaApi types (which are only partially unrelated, see DataPlaneOverviewParameters). Let me know if you want me to move out the StatusBadge changes.

I think we are good now, I've been looking over the other PRs split off for this, and judging by those this should be a bit smaller now. I'll take another look right now though and let you know.

kleinfreund commented:
**Question**: Should the DPP table for services show gateways? If so, should that table include both protocol (DPP specific and type (gateway specific) columns>

johncowen commented on src/types/index.d.ts:
```diff
@@ -44,6 +44,11 @@ export type TableData = {
+  title: 'Online' | 'Offline' | 'Partially degraded' | 'Not available'
+  appearance: 'warning' | 'danger' | 'success'
+}
+
```
We spoke about this a little offline, just to note that it would be nice to keep the English text and the styling/visuals out of the data and use something like an i18n lib and S/CSS for styling/visuals.

```javascript
// but dynamically calculating the Status. i.e. `status` here would come out as `not_available`
$t(`service.status.${status}`)
// and
:class="`status-${status}`"
```

For me at the very least taking this approach means less code to write and follow and keeps everything in the one place in the component.

Overall, right here right now, not a blocker though, just wanted to note it here.

kleinfreund replied:
Have a look at the changes I added in https://github.com/kumahq/kuma-gui/pull/479/commits/c5379f44bdc52d87749dca1db3bbb9f5eee24fdb.

This isolates the mapping of "status keywords" in the status badge component and simplifies consuming code to only pass around the status keyword as it already is present on the respective data strucutres.

---

One thing I really don’t agree with is the analogy of this sort of data transformation being akin to internationalization/localization. It isn’t. We’re transforming some machine-readable data into a very few select human-readable labels and machine-readable props for the use with Kongponents. This doesn’t justify the use of an i18n library.

johncowen replied:
Much nicer, I like that this stuff is contained in the component. Just to note: Not saying we need to pull in an i18n right here right now, but it feels to me like we'll need one at some point to do machine > human translations rather than just rolling our own (probably slightly differently each time) translation for each individual instance of this sort of stuff. Anyway one to chat about at a later date.

johncowen commented on src/router/router.ts:
```diff
@@ -102,11 +102,11 @@ export function createRouter(baseGuiPath: string = '/', policyDefinitions: Polic
               props: (route) => ({
                 name: route.query.name,
                 offset: getLastNumberParameter(route.query.offset),
+                isGatewayView: true,
```
As mentioned offline, much prefer using `props` for this, weirdly I couldn't seem to get it working with props on my first attempt, I must have been doing something wrong!

johncowen reviewed (approved):
Thanks for walking me through this! I left a few comments, but mainly around what we briefly discussed, but otherwise LGTM!

kleinfreund commented:
> If we do need to add them, if it was me for the minute I'd vote for just adding a column for the moment, but maybe consider the tabbed interface later.

I think combining both in the same tables is the wrong approach: we *just* split off gateways from data plane proxies because they’re different things. We should follow this thought and if we want to show service-attached gateways in the service detail view, they should be shown separately from data plane proxies.

In the interest of not blowing up the scope of this task, I propose to make this a follow-up task and ensure that we only show DPPs and not gateways in the service detail view for now.

kleinfreund commented on src/app/data-planes/components/DataPlaneEntitySummary.vue:
```diff
@@ -1,146 +1,145 @@
+                class="ml-1"
+                icon="warning"
+                color="var(--black-75)"
+                secondary-color="var(--yellow-300)"
```
We decided to keep the Kongponents yellow.

kleinfreund commented:
I’ve filed https://github.com/kumahq/kuma-gui/issues/502 following https://github.com/kumahq/kuma-gui/pull/479#issuecomment-1347993632.

johncowen commented:
> I think combining both in the same tables is the wrong approach: we just split off gateways from data plane proxies because they’re different things. We should follow this thought and if we want to show service-attached gateways in the service detail view, they should be shown separately from data plane proxies.

Agree. Ideally the tabbed thing was preferable, and the column was short term, but sounds like even the column isn't short term. So yeah, follow up task for the tabbed interface here if thats what we want and move on, thats sounds good to me!
